### PR TITLE
Update debugging doc

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Thanks, @ananthsub. Missed it. BTW anything else do you think is worth adding?
https://github.com/PyTorchLightning/pytorch-lightning/pull/11390#pullrequestreview-850729055

Changes:
- Added detect anomaly and DDP Debugging
- Move things around to group them.

Part of docs revamp #7666

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @rohitgr7

### CODE REVIEW

**Review by ananthsub — COMMENTED** [review]
@rohitgr7 overall, I'm not really sure what the Debugging section is supposed to answer. The features here are all slightly separate.

If I tried grouping them into categories, it could be:
- Fail fast: fast_dev_run, limit_train/val/test/predict batches, overfit batches, sanity check
- Model observability: Model Summary
- Performance debugging: Profiling, device stats
- Optimization debugging: detect anomaly, track grad norm

**Comment by rohitgr7** [comment]
@ananthsub 

just an attempt to provide users a page where they can go for debugging their models. It can be anything, bottlenecks in training code, unit testing, quick prototyping. Otherwise, they had to go and look for a complete trainer page or other pages and check which flag might help them.

Yes, I realized that they are not quite aligned together after reading it myself and grouping them already. Still WIP.

**Review by akihironitta — COMMENTED** [review]
How about adding a section for debugging ddp in a cpu environment, too, as mentioned in the v1.5.0 release note: https://github.com/PyTorchLightning/pytorch-lightning/releases/1.5.0

<details>

> ## DDP Debugging Improvements
> Are you having a hard time debugging DDP on your remote machine? Now you can debug DDP locally on the CPU:
> ```python 
> trainer = Trainer(accelerator="cpu", strategy="ddp", devices=2)
> ```
> When everything works, switch back to GPU by changing only the `accelerator`. Check our documentation for more [useful debugging tricks](https://pytorch-lightning.readthedocs.io/en/1.5.0/common/debugging.html).
> Note that this will not provide any speed benefits.

</details>

**Comment by akihironitta** [comment]
@rohitgr7 Docs build seems failing. Could you have a look?

File: docs/source/common/debugging.rst [hunk 1]
```diff
@@ -97,8 +106,73 @@ argument of :class:`~pytorch_lightning.trainer.trainer.Trainer`)
...
+
+
+************
+Optimization
+************
+
+Inspect Gradient Norms
+======================
+
+Logs (to a logger), the norm of each weight matrix.
```
Comment (awaelchli) [inline docs/source/common/debugging.rst:120]: No, it's not the weight matrix at all. This is taking the norm of the gradients. and it is not a matrix in general.

File: docs/source/common/debugging.rst [hunk 2]
```diff
@@ -97,8 +106,73 @@ argument of :class:`~pytorch_lightning.trainer.trainer.Trainer`)
...
+
+----------------
+
+
+***********
+Performance
+***********
+
+Log Device Stats
+================
```
Comment (awaelchli) [inline docs/source/common/debugging.rst:152]:
```suggestion
Log Device Statistics
=====================
```

File: docs/source/common/debugging.rst [hunk 3]
```diff
@@ -133,43 +207,25 @@ See Also:
...
+Debugging with Distributed Strategies
+*************************************
 
-(See: :paramref:`~pytorch_lightning.trainer.trainer.Trainer.num_sanity_val_steps`
-argument of :class:`~pytorch_lightning.trainer.trainer.Trainer`)
+DDP Debugging
+=============
 
-.. testcode::
+If you are having a hard time debugging DDP on your remote machinem you can debug DDP locally on the CPU:
```
Comment (awaelchli) [inline docs/source/common/debugging.rst:221]:
```suggestion
If you are having a hard time debugging DDP on your remote machine you can debug DDP locally on the CPU:
```

File: docs/source/common/debugging.rst [hunk 4]
```diff
@@ -133,43 +207,25 @@ See Also:
...
-    # DEFAULT
-    trainer = Trainer(num_sanity_val_steps=2)
+.. code-block:: python
 
-----------------
+    trainer = Trainer(accelerator="cpu", strategy="ddp", devices=2)
 
-Profiling
-=========
+When everything works, switch back to GPU by changing only the accelerator. Note that this will not provide any speed benefits.
```
Comment (awaelchli) [inline docs/source/common/debugging.rst:227]:
```suggestion
When everything works, switch back to GPU by changing only the accelerator.
```
This cannot be mentioned here. "This" refers to the previous sentence suggesting to switch back to GPU, and this indeed should generally provide speed benefits!!!

**Review by awaelchli — APPROVED** [review]

File: docs/source/common/debugging.rst (hunk 4 above)
Comment (rohitgr7) [reply docs/source/common/debugging.rst:227]: yeah, it was just referring to the overall trick. But yeah, might be confusing. Updated!

File: docs/source/common/debugging.rst (hunk 1 above)
Comment (rohitgr7) [reply docs/source/common/debugging.rst:120]: good catch!

File: docs/source/common/debugging.rst [hunk 5]
```diff
@@ -133,43 +207,25 @@ See Also:
...
+Debugging with Distributed Strategies
+*************************************
 
-(See: :paramref:`~pytorch_lightning.trainer.trainer.Trainer.num_sanity_val_steps`
-argument of :class:`~pytorch_lightning.trainer.trainer.Trainer`)
+DDP Debugging
+=============
 
-.. testcode::
+If you are having a hard time debugging DDP on your remote machine you can debug DDP locally on the CPU. Note that this will not provide any speed benifits.
```
Comment (tchaton) [inline docs/source/common/debugging.rst:221]: I would add hints on how to debug with DDP.
First, debug with printing the ranks print(f"RANK: {trainer.global_rank} {msg}"). IDE supports exploring multi process such as DDP SPawn.
Futhermore, you can use use breakpoint by protecting it with rank information.
```
if trainer.global_rank == 9:
      breakpoint()
...
```

Comment (tchaton) [reply docs/source/common/debugging.rst:221]: I would really like to have _rank_breakpoint(0) and multi_print(msg) (using color) to help debugging DDP. Maybe next PRs :)

Comment (rohitgr7) [reply docs/source/common/debugging.rst:221]: it often hangs a lot when I do this with DDP since we can't stop other processes from executing further. Curious to see how rank_breakpoint will work.

**Review by justusschock — APPROVED** [review]

File: docs/source/common/debugging.rst (hunk 5 above)
Comment (awaelchli) [reply docs/source/common/debugging.rst:221]: a rank breakpoint could be implemented like this (did not test):

```py
def rank_breakpoint(rank=0):
    if global_rank != rank:
        barrier()  # all processes wait here except the one we want to debug
    else:
        # the chosen rank will set a trace, but we want the frame from where this function is called
        pdb.Pdb().set_trace(inspect.currentframe().f_back.f_back)
        # once breakpoint gets skipped, synchronize with all other workers
        barrier()
```

File: docs/source/common/debugging.rst [hunk 6]
```diff
@@ -133,43 +207,44 @@ See Also:
...
 
-Lightning runs a few steps of validation in the beginning of training.
-This avoids crashing in the validation loop sometime deep into a lengthy training loop.
+    class LitModel(LightningModule):
+        def training_step(self, batch, batch_idx):
 
-(See: :paramref:`~pytorch_lightning.trainer.trainer.Trainer.num_sanity_val_steps`
-argument of :class:`~pytorch_lightning.trainer.trainer.Trainer`)
+            debugging_message = ...
+            print(f"RANK - {self.trainer.local_rank}: {debugging_message}")
```
Comment (rohitgr7) [inline docs/source/common/debugging.rst:236]:
```suggestion
            print(f"RANK - {self.trainer.global_rank}: {debugging_message}")
```

File: docs/source/common/debugging.rst [hunk 7]
```diff
@@ -133,43 +207,44 @@ See Also:
...
 
-.. testcode::
+            if self.trainer.local_rank == 0:
```
Comment (rohitgr7) [inline docs/source/common/debugging.rst:238]:
```suggestion
            if self.trainer.global_rank == 0:
```

File: docs/source/common/debugging.rst (hunk 5 above)
Comment (rohitgr7) [reply docs/source/common/debugging.rst:221]: thanks @awaelchli 
no-barrier was the reason why it used to hang a lot for me with DDP + pdb

I think we can add this.. once tested.

File: docs/source/common/debugging.rst [hunk 8]
```diff
@@ -133,43 +207,44 @@ See Also:
...
-    trainer = Trainer(limit_train_batches=10, limit_val_batches=5)
+    trainer = Trainer(accelerator="cpu", strategy="ddp", devices=2)
 
-----------------
+To inspect your code, you can use `pdb <https://docs.python.org/3/library/pdb.html>`_ or `breakpoint() <https://docs.python.org/3/library/functions.html#breakpoint>`_
+or use regular print statements.
 
-Validation Sanity Check
-=======================
+.. code-block:: python
```
Comment (Borda) [inline docs/source/common/debugging.rst:230]: lets use test code as much as possible to about syntax errors and API missalignment

File: docs/source/common/debugging.rst [hunk 9]
```diff
@@ -47,35 +53,38 @@ argument of :class:`~pytorch_lightning.trainer.trainer.Trainer`)
...
 
-----------------
 
-Inspect Gradient Norms
-======================
+Shorten Epochs
+==============
 
-Logs (to a logger), the norm of each weight matrix.
+Sometimes it's helpful to only use a percentage of your training, val, test or predict data (or a set number of batches).
```
Comment (Borda) [inline docs/source/common/debugging.rst:60]: I may say rather that in most cases it is sufficient to use just a fraction, up to 100 batches

File: docs/source/common/debugging.rst [hunk 10]
```diff
@@ -97,8 +106,73 @@ argument of :class:`~pytorch_lightning.trainer.trainer.Trainer`)
...
+
+
+************
+Optimization
+************
+
+Inspect Gradient Norms
+======================
+
+Logs (to a logger), the norm of the gradients.
```
Comment (Borda) [inline docs/source/common/debugging.rst:120]: it is a full sentence, missing verb... :)

File: docs/source/common/debugging.rst [hunk 11]
```diff
@@ -97,8 +106,73 @@ argument of :class:`~pytorch_lightning.trainer.trainer.Trainer`)
...
+    trainer = Trainer(track_grad_norm=2)
+
+
+Detect Anomaly
+==============
+
+You can enable anomaly detection for the autograd engine. It uses PyTorch's built-in
+`Anomaly Detection Context-manager <https://pytorch.org/docs/stable/autograd.html#anomaly-detection>`_.
+
+To enable it within Lightning, use:
```
Comment (Borda) [inline docs/source/common/debugging.rst:137]:
```suggestion
To enable it within Lightning, use Trainer's flag:
```

**Review by Borda — APPROVED** [review]

File: docs/source/common/debugging.rst [hunk 12]
```diff
@@ -133,43 +207,44 @@ See Also:
...
+*************************************
+Debugging with Distributed Strategies
+*************************************
 
-On larger datasets like Imagenet, this can help you debug or test a few things faster than waiting for a full epoch.
+DDP Debugging
+=============
 
-.. testcode::
+If you are having a hard time debugging DDP on your remote machine you can debug DDP locally on the CPU. Note that this will not provide any speed benifits.
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply docs/source/common/debugging.rst:221]: this covers independent researchers/practitioners.. we just want to highlight the solutions. Problem can be anything 😄

File: docs/source/common/debugging.rst (hunk 9 above)
Comment (rohitgr7) [inline docs/source/common/debugging.rst:60]:
```suggestion
Sometimes it's helpful to only use a fraction of your training, val, test, or predict data (or a set number of batches).
```

File: docs/source/common/debugging.rst (hunk 10 above)
Comment (rohitgr7) [inline docs/source/common/debugging.rst:120]:
```suggestion
Logs the norm of the gradients to the logger.
```
